### PR TITLE
Add structure to support class structures as type

### PR DESCRIPTION
**Pull request by theEvilReaper**
## Proposed changes

Some of the spec structures require a `String` value to represent the type of a value. While using a String value is acceptable, it has the potential for more errors, such as misspelling a type or something else. A better approach is to have a way to accept the ClassName from a given `Class`. This allows the usage of `String::class`, which is safer than using a string value.

The pull request introduces two implementation of a TypeName:

`TypeName` represents only the name of a class
`ClassName` also represents a class but can have the option to be nullable
`ParameterizedTypeName`  can be used to combine a typename with parameterized objects.

## Types of changes

What types of changes does your code introduce to this project?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of
them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before
merging your code._

- [x] I have read the CONTRIBUTING.md
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you
did and what alternatives you considered, etc...

### CODE REVIEW

File: src/main/kotlin/net/theevilreaper/dartpoet/code/writer/FunctionWriter.kt [hunk 1]
```diff
@@ -28,7 +28,9 @@ class FunctionWriter {
             }
         }
 
-        if (functionSpec.returnType.orEmpty().trim().isEmpty()) {
+
+
+        if (functionSpec.returnType == null) {
```
[the remark this answers is not in the document]
Comment (theEvilReaper) [reply src/main/kotlin/net/theevilreaper/dartpoet/code/writer/FunctionWriter.kt:33]: The returnType value is a `TypeName` which doesn't have the empty check

File: src/main/kotlin/net/theevilreaper/dartpoet/type/TypeName.kt [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+@JvmField
+val DYNAMIC: ClassName = DynamicClassName()
+
+@JvmName("get")
+fun KClass<*>.asTypeName(): TypeName = asClassName()
+
+/** Returns a [TypeName] equivalent to this [Type].  */
+@JvmName("get")
+fun Type.asTypeName(): TypeName = TypeName.get(this)
```
[the remark this answers is not in the document]
Comment (theEvilReaper) [reply src/main/kotlin/net/theevilreaper/dartpoet/type/TypeName.kt:177]: It's a bit weird but the method calls the get method from the `TypeName` companion object

[the remark this answers is not in the document]
Comment (TheMeinerLP) [reply src/main/kotlin/net/theevilreaper/dartpoet/type/TypeName.kt:177]: I prefer for that a extra class/file to have it more organized

[the remark this answers is not in the document]
Comment (theEvilReaper) [reply src/main/kotlin/net/theevilreaper/dartpoet/type/TypeName.kt:177]: I can outsource all constant values with the 'JvmName' annotation in another class, but the other values can't be moved

**Review by TheMeinerLP — APPROVED** [review]

**Pull request merged**